private/pkg/manifest/manifest.go

```diff
@@ -0,0 +1,269 @@
+	msg    string
+}
+
+var _ error = (*ManifestError)(nil)
+
+func newManifestError(lineno int, msg string) *ManifestError {
+	return &ManifestError{
+		lineno: lineno,
+		msg:    msg,
+	}
+}
+
+func (e *ManifestError) Error() string {
+	return fmt.Sprintf("invalid manifest: %d: %s", e.lineno, e.msg)
+}
+
+// Manifest represents a list of paths and their digests.
+type Manifest struct {
+	paths   map[string]*Digest
+	digests map[string]string
+	hash    sha3.ShakeHash
+}
+
+var _ encoding.TextMarshaler = (*Manifest)(nil)
+var _ encoding.TextUnmarshaler = (*Manifest)(nil)
+
+// NewManifest creates an empty manifest.
+func NewManifest() *Manifest {
+	return &Manifest{
+		paths:   make(map[string]*Digest),
+		digests: make(map[string]string),
+		hash:    sha3.NewShake256(),
+	}
+}
+
+func splitManifest(data []byte, atEOF bool) (int, []byte, error) {
+	// Return a line without LF.
+	if i := bytes.IndexByte(data, '\n'); i >= 0 {
+		return i + 1, data[0:i], nil
+	}
+
+	// EOF occurred with a partial line.
+	if atEOF && len(data) != 0 {
+		return 0, nil, errPartial
+	}
+
+	return 0, nil, nil
+}
+
+// NewManifestFromReader builds a manifest an encoded manifest reader.
+func NewManifestFromReader(manifest io.Reader) (*Manifest, error) {
+	m := NewManifest()
+	scanner := bufio.NewScanner(manifest)
+	scanner.Split(splitManifest)
+	lineno := 0
+	for scanner.Scan() {
+		lineno++
+		hash, path, found := strings.Cut(scanner.Text(), "  ")
+		if !found {
+			return nil, newManifestError(lineno, "invalid entry")
+		}
+		digest, err := NewDigestFromString(hash)
+		if err != nil {
+			return nil, multierr.Append(
+				newManifestError(lineno, "invalid hash"),
+				err,
+			)
+		}
+		if digest.Type() != shake256Name {
+			return nil, newManifestError(lineno, "unknown hash")
+		}
+		if len(digest.Bytes()) != 64 {
+			return nil, newManifestError(lineno, "short digest")
+		}
+		m.addDigest(path, digest)
+	}
+	err := scanner.Err()
+	if err == errPartial {
+		return nil, newManifestError(lineno, "partial record")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// NewManifestFromBucket creates a manifest from a bucket.
+func NewManifestFromBucket(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+) (*Manifest, error) {
+	m := NewManifest()
+	err := bucket.Walk(ctx, "", func(info storage.ObjectInfo) error {
+		path := info.Path()
+		obj, err := bucket.Get(ctx, path)
+		if err != nil {
+			return err
+		}
+		m.AddContent(path, obj)
+		return obj.Close()
+	})
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (m *Manifest) addDigest(path string, digest *Digest) {
+	m.paths[path] = digest
+	m.digests[digest.String()] = path
+}
+
+// AddContent adds a manifest entry for path by its content.
+func (m *Manifest) AddContent(path string, content io.Reader) {
+	m.hash.Reset()
+	// sha3.ShakeHash never errors, reading or writing. These checks are to
+	// satisfy linting, possibly exploding if some breaking behavior change
+	// happens.
```

Comment:
I think we might just want to have this return an error (even if unlikely). That is probably preferred to a panic coming from lower level code.

Reply:
This one is a bit frustrating because the hashing should never error. You'll see this don't-check-error even in Go's crypto libraries (e.g. [sha2](https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/crypto/sha256/sha256.go;l=263-285), [sha3](https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.1.0:sha3/hashes.go;l=67-97)).

However, keeping content in memory sucks too. Propagating any read error from `content` is useful.